core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -283,7 +283,10 @@ public void commit() {
             Snapshot newSnapshot = apply();
             newSnapshotId.set(newSnapshot.snapshotId());
             TableMetadata updated;
-            if (stageOnly) {
+            if (base.snapshot(newSnapshot.snapshotId()) != null) {
+              // this is a rollback operation
+              updated = base.replaceCurrentSnapshot(newSnapshot.snapshotId());
+            } else if (stageOnly) {
```

Comment:
Can you port the changes from the other fix? That removed methods from `TableMetadata` rather than adding new ones. I think it would be best to keep 0.13.x and master in sync rather than having slightly different fixes.

Thanks, @sririshindra!

Reply:
@rdblue, would the following be acceptable? (from line 285)
```
            TableMetadata.Builder update = TableMetadata.buildFrom(base);
            if (base.snapshot(newSnapshot.snapshotId()) != null) {
              // this is a rollback or cherrypick operation
              update.setCurrentSnapshot(newSnapshot.snapshotId());
              } else if (stageOnly) {
              update.addSnapshot(newSnapshot);
            } else {
              update.setCurrentSnapshot(newSnapshot);
            }

            TableMetadata updated = update.build();
            if (updated.changes().isEmpty()) {
              // do not commit if the metadata has not changed. for example, this may happen when setting the current
              // snapshot to an ID that is already current. note that this check uses identity.
              return;
            }
```
I know the point of #4089 is to replace `setCurrentSnapshot` with `setBranchSnapshot`, but in 0.13.x, we don't support branching yet, and the commits #4089 builds on are not present.

Reply:
Yes, calling the right `setCurrentSnapshot` is a good way to backport this. Thanks!